flink-ml-core/src/main/java/org/apache/flink/ml/common/iteration/TerminateOnMaxIterOrTol.java

```diff
@@ -16,29 +16,53 @@
  * limitations under the License.
  */
 
-package org.apache.flink.test.iteration.operators;
+package org.apache.flink.ml.common.iteration;
 
 import org.apache.flink.api.common.functions.FlatMapFunction;
 import org.apache.flink.iteration.IterationListener;
 import org.apache.flink.util.Collector;
 
-/** An termination criteria function that asks to stop after the specialized round. */
-public class RoundBasedTerminationCriteria
-        implements FlatMapFunction<EpochRecord, Integer>, IterationListener<Integer> {
+/**
+ * A FlatMapFunction that emits values iff the iteration's epochWatermark does not exceed a certain
+ * threshold and the loss exceeds a certain tolerance.
+ *
+ * <p>When the output of this FlatMapFunction is used as the termination criteria of an iteration
+ * body, the iteration will be executed for at most the given `maxIter` iterations. And the
+ * iteration will terminate once any input value is smaller than or equal to the given `tol`.
+ */
+public class TerminateOnMaxIterOrTol
+        implements IterationListener<Integer>, FlatMapFunction<Double, Integer> {
+
+    private final int maxIter;
 
-    private final int maxRound;
+    private final double tol;
 
-    public RoundBasedTerminationCriteria(int maxRound) {
-        this.maxRound = maxRound;
+    private double loss = Double.NEGATIVE_INFINITY;
```

Comment:
What happens if there are not input values in the first epoch? Do we expect the iteration to terminate (which seems to be the case with `loss = Double.NEGATIVE_INFINITY`)?

Reply:
Ah, thanks for pointing this out. It should be `Double.MAX_VALUE`

Reply:
There should always be one loss value in each epoch (including the first one).

Thanks for pointing this out, I think I made a mistake here --- the initial value of `loss` should be `Double.MAX_VALUE`.

Reply:
Sounds good. Thanks.